apache/zeppelin
# ZEPPELIN-539 RemoteInterpreter Heartbeat (WIP)

### What is this PR for?

To help users to determine remote interpreter is not able to respond.
This is just WIP, and "How to help users" could be improved with discussions.

### What type of PR is it?

Feature (Improve?)

### Todos
* [x] - Rebase when #574 is merged to master
  * since shutdowning heartbeat threads requires reference count to be zero
  * without #574 some threads could be alive although remote interpreter is closed
* [ ] - Discuss proper values for sending heartbeat interval, checking timeout interval
* [ ] - Discuss how to let users know when remote interpreter is timed out
* [ ] - Discuss possible way to restore remote interpreter back to normal

### Is there a relevant Jira issue?

https://issues.apache.org/jira/browse/ZEPPELIN-539

### How should this be tested?

1. run a spark paragraph to ensure spark remote interpreter process is run
2. kill -9 to spark remote interpreter process
3. run paragraph again (it may show broken pipe, or connection refused after #575)
4. wait 30 secs (or remote interpreter connection timeout value) to let RemoteInterpreterProcess classifies process to be timed out
5. run paragraph again (it shows org.apache.zeppelin.interpreter.InterpreterProcessHeartbeatFailedException to users)

### Screenshots (if appropriate)

<img width="1081" alt="2015-12-29 7 34 04" src="https://cloud.githubusercontent.com/assets/1317309/12026576/9cc40d6c-adfe-11e5-9207-09e1f019f251.png">

### Questions:
* Does the licenses files need update? (No)
* Is there breaking changes for older versions? (No)
* Does this needs documentation? (Maybe no)
